### PR TITLE
Fix: RPC parallel test runs

### PR DESCRIPTION
**Pull request by sudo-shashank**
## Summary of changes

Changes introduced in this pull request:

- Added `max_concurrent_requests` to limit max concurrent test runs which prevents overloading the rpc servers.
This issue was observed for `lotus` only and not for `forest`, when approx 1000+ RPC tests were executed simultaneously.

## Reference issue to close (if applicable)

Closes

## Change checklist

- [X] I have performed a self-review of my own code,
- [X] I have made corresponding changes to the documentation,
- [X] I have added tests that prove my fix is effective or that my feature works (if possible),
- [X] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
> Added max_concurrent_requests to limit max concurrent test runs which prevents overloading the rpc servers

Could you please elaborate on this? Which conditions made the RPC servers overload, at what QPS? Was it Lotus or Forest? It's useful to know these limitations.

**Comment by sudo-shashank** [comment]
> > Added max_concurrent_requests to limit max concurrent test runs which prevents overloading the rpc servers
> 
> Could you please elaborate on this? Which conditions made the RPC servers overload, at what QPS? Was it Lotus or Forest? It's useful to know these limitations.

updated PR description

**Comment by LesnyRumcajs** [comment]
Thanks, this is interesting. I believe we should benchmark Forest and Lotus for the Top N queries, we might need it for hardware/replicas count recommendations.

File: src/tool/subcommands/api_cmd.rs [hunk 1]
```diff
@@ -50,9 +51,20 @@ pub enum ApiCommands {
         #[arg(long, value_enum, default_value_t = RunIgnored::Default)]
         /// Behavior for tests marked as `ignored`.
         run_ignored: RunIgnored,
+        /// Maximum number of concurrent requests
+        #[arg(long, default_value = "8")]
+        max_concurrent_requests: usize,
     },
 }
 
+struct ApiTestFlags {
```
Comment (LesnyRumcajs) [inline src/tool/subcommands/api_cmd.rs:60]: Let's document this. I have no clue what, for example, a `String` `filter` is and what does it expect.

**Comment by elmattic** [comment]
Is this a problem of API calls for Lotus being rate-limited by default?

We could run Lotus without limitation if this is the case.

File: src/tool/subcommands/api_cmd.rs [hunk 2]
```diff
@@ -632,41 +641,43 @@ async fn compare_apis(
...
 async fn run_tests(
     tests: Vec<RpcTest>,
     forest: &ApiInfo,
     lotus: &ApiInfo,
-    filter: String,
-    fail_fast: bool,
-    run_ignored: RunIgnored,
+    config: &ApiTestFlags,
 ) -> anyhow::Result<()> {
+    let semaphore = Arc::new(Semaphore::new(config.max_concurrent_requests));
```
[the remark this answers is not in the document]
Comment (elmattic) [reply src/tool/subcommands/api_cmd.rs:658]: We have some semaphores in the `RaceBatch`  😉

File: src/tool/subcommands/api_cmd.rs [hunk 3]
```diff
@@ -50,9 +51,20 @@ pub enum ApiCommands {
         #[arg(long, value_enum, default_value_t = RunIgnored::Default)]
         /// Behavior for tests marked as `ignored`.
         run_ignored: RunIgnored,
+        /// Maximum number of concurrent requests
+        #[arg(long, default_value = "8")]
+        max_concurrent_requests: usize,
```
Comment (elmattic) [inline src/tool/subcommands/api_cmd.rs:56]: Supporting a way to turn off concurrent requests limitation could be nice (i.e., using the exceptional value 0). But for now I think this is ok.

File: src/tool/subcommands/api_cmd.rs (hunk 1 above)
Comment (sudo-shashank) [reply src/tool/subcommands/api_cmd.rs:60]: we already have a short description for these flags under `ApiCommands`
https://github.com/ChainSafe/forest/blob/ced0f7570ccc9e52b01db3ddf56dc0439288da91/src/tool/subcommands/api_cmd.rs#L41-L52

Comment (LesnyRumcajs) [reply src/tool/subcommands/api_cmd.rs:60]: I wasn't aware of that. Perhaps adding a Rust doc link to the `ApiCommands` struct would help future readers. What do you think?

**Review by LesnyRumcajs — APPROVED** [review]

File: src/tool/subcommands/api_cmd.rs (hunk 1 above)
Comment (sudo-shashank) [reply src/tool/subcommands/api_cmd.rs:60]: yes, i'll add

**Comment by elmattic** [comment]
> This issue was observed for lotus only and not for forest

Looking at trace logs, I can see lots of errors for Forest as well if I use "unlimited" mode (i.e. `--max-concurrent-requests 10000`):

`error sending request for URL (http://127.0.0.1:2345/rpc/v0): connection error: Connection reset by peer (os error 54)`

`2345` is the port used by Forest.

Btw, I fixed some because I had to low FDs (so using `ulimit -n 8192`).

**Review by elmattic — APPROVED** [review]

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**